### ISSUE TITLE
Validity of ManuallyDrop, or: ManuallyDrop<Box<_>> is strange

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/unsafe-code-guidelines#245]
The `ManuallyDrop` docs say

> ManuallyDrop<T> is subject to the same layout optimizations as T. As a consequence, it has no effect on the assumptions that the compiler makes about its contents. 

In particular, this means that a `ManuallyDrop<Box<T>>` must not dangle. But that means that after calling `drop` on a `ManuallDrop<Box<T>>`, that value violates the validity invariant and must not be used or passed around any more! I don't think that is the behavior we want.

We want `ManuallyDrop` to preserve niches, so we cannot really relax what it says about validity. But what we could relax is validity of `Box` -- we could say that validity is purely about the pointer value itself (it must be non-NULL and properly aligned). We could specify that the "dereferencability" only comes in through the aliasing model, and we could make that model stop looking into `ManuallyDrop`.

**Comment by theemathas** [comment]
Of note, `ManuallyDrop<&mut T>` has the same behavior, although it's harder to accidentally run into. For example, the Miri says the following code has UB.
```rust
use std::mem::{ManuallyDrop, transmute};

fn main() {
    let mut x = Box::new(123);
    let y: &mut i32 = unsafe { transmute(&mut *x) };
    let mut z = ManuallyDrop::new(y);
    // Removing the following line does not affect the behavior
    unsafe { ManuallyDrop::drop(&mut z); }
    drop(x);
    let _moved = z;
}
```

**Comment by RalfJung** [comment]
Yeah, but it's less strange there since dropping `&mut T` is a NOP anway. So this issue focuses on `ManuallyDrop<Box<T>>`; `&mut` is just in it for consistency reasons.

**Comment by RalfJung** [comment]
Closing in favor of https://github.com/rust-lang/rust/issues/118166
